pydantic_ai_slim/pydantic_ai/toolsets/duckduckgo.py

```diff
@@ -0,0 +1,53 @@
+from dataclasses import dataclass
+from typing import TypedDict
+
+import anyio
+import anyio.to_thread
+from pydantic import TypeAdapter
+
+from pydantic_ai.tools import RunContext, Tool
+
+try:
+    from duckduckgo_search import DDGS
+except ImportError as _import_error:
+    raise ImportError(
+        'Please install `duckduckgo-search` to use the DuckDuckGo search tool, '
+        "you can use the `duckduckgo` optional group — `pip install 'pydantic-ai-slim[duckduckgo]'`"
+    ) from _import_error
+
+
+class DuckDuckGoResult(TypedDict):
+    """A DuckDuckGo search result."""
+
+    title: str
+    """The title of the search result."""
+    href: str
+    """The URL of the search result."""
+    body: str
+    """The body of the search result."""
+
+
+duckduckgo_ta = TypeAdapter(list[DuckDuckGoResult])
+
+
+@dataclass
+class DuckDuckGoSearchTool:
+    """The DuckDuckGo search tool."""
+
+    client: DDGS
+    """The DuckDuckGo search client."""
+
+    async def __call__(self, ctx: RunContext, query: str) -> list[DuckDuckGoResult]:
+        results = await anyio.to_thread.run_sync(self.client.text, query)
+        if len(results) == 0:
+            raise RuntimeError('No search results found.')
+        return duckduckgo_ta.validate_python(results)
+
+
+def duckduckgo_search_tool(duckduckgo_client: DDGS | None = None):
```

Comment:
I feel like this file needs a `__all__` that only contains `'duckduckgo_search_tool'`, or the other stuff in it should be made leading-`_` private. Unless you intend for users to use the other classes etc., but it doesn't seem like it?

Reply:
Only PyCharm users care about `__all__`... I'll add it.